### PR TITLE
8283232: x86: Improve vector broadcast operations

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch improves the generation of broadcasting a scalar in several ways:

- As it has been pointed out, dumping the whole vector into the constant table is costly in terms of code size, this patch minimises this overhead for vector replicate of constants. Also, options are available for constants to be generated with more alignment so that vector load can be made efficiently without crossing cache lines.
- Vector broadcasting should prefer rematerialising to spilling when register pressure is high.
- Load vectors using the same kind (integral vs floating point) of instructions as that of the results to avoid potential data bypass delay

With this patch, the result of the added benchmark, which performs some operations with a really high register pressure, on my machine with Intel i7-7700HQ (avx2) is as follow:

                                              Before          After
    Benchmark                  Mode  Cnt   Score   Error   Score   Error  Units     Gain
    SpiltReplicate.testDouble  avgt    5  42.621 ± 0.598  38.771 ± 0.797  ns/op   +9.03%
    SpiltReplicate.testFloat   avgt    5  42.245 ± 1.464  38.603 ± 0.367  ns/op   +8.62%
    SpiltReplicate.testInt     avgt    5  20.581 ± 5.791  13.755 ± 0.375  ns/op  +33.17%
    SpiltReplicate.testLong    avgt    5  17.794 ± 4.781  13.663 ± 0.387  ns/op  +23.22%

As expected, the constant table sizes shrink significantly from 1024 bytes to 256 bytes for `long`/`double` and 128 bytes for `int`/`float` cases.

This patch also removes some redundant code paths and renames some incorrectly named instructions.

Thank you very much.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8283232](https://bugs.openjdk.org/browse/JDK-8283232): x86: Improve vector broadcast operations

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/7832/head:pull/7832` \
`$ git checkout pull/7832`

Update a local copy of the PR: \
`$ git checkout pull/7832` \
`$ git pull https://git.openjdk.org/jdk pull/7832/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7832`

View PR using the GUI difftool: \
`$ git pr show -t 7832`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/7832.diff">https://git.openjdk.org/jdk/pull/7832.diff</a>

</details>

### CODE REVIEW

**Comment by jatin-bhateja** [comment]
Will be helpful  is a JMH can be created around this, following is the except from  X86 Optimizations manual section Appendix E Section E.1.3 
"Forwarding the result within the same bypass domain from a producer micro-op to a consumer micro is done efficiently in hardware without delay"

**Comment by merykitty** [comment]
Hi, forwarding results within the same bypass domain does not result in delay, data bypass delay happens when the data crosses different domains, according to "Intel® 64 and IA-32 Architectures Optimization Reference Manual"

> When a source of a micro-op executed in one stack comes from a micro-op executed in another stack, a delay can occur. The delay occurs also for transitions between Intel SSE integer and Intel SSE floating-point operations. In some of the cases, the data transition is done using a micro-op that is added to the instruction flow.

The manual mentions the guideline at section 3.5.2.2

image

Thanks.

**Comment by jatin-bhateja** [comment]
> Hi, forwarding results within the same bypass domain does not result in delay, data bypass delay happens when the data crosses different domains, according to "Intel® 64 and IA-32 Architectures Optimization Reference Manual"
> 
> > When a source of a micro-op executed in one stack comes from a micro-op executed in another stack, a delay can occur. The delay occurs also for transitions between Intel SSE integer and Intel SSE floating-point operations. In some of the cases, the data transition is done using a micro-op that is added to the instruction flow.
> 
> The manual mentions the guideline at section 3.5.2.2
> 
> image
> 
> Thanks.

Thanks meant to refer to above text.  I have removed incorrect reference.

**Comment by jatin-bhateja** [comment]
> > Hi, forwarding results within the same bypass domain does not result in delay, data bypass delay happens when the data crosses different domains, according to "Intel® 64 and IA-32 Architectures Optimization Reference Manual"
> > > When a source of a micro-op executed in one stack comes from a micro-op executed in another stack, a delay can occur. The delay occurs also for transitions between Intel SSE integer and Intel SSE floating-point operations. In some of the cases, the data transition is done using a micro-op that is added to the instruction flow.
> > 
> > 
> > The manual mentions the guideline at section 3.5.2.2
> > image
> > Thanks.
> 
> Thanks meant to refer to above text. I have removed incorrect reference.

It will still be good if we can come up with a micro benchmark, that shows the gain with the patch.

**Comment by merykitty** [comment]
Doing a simple benchmark that has a lot of register pressure

    @Benchmark
    public long broadcastCon() {
        var species = IntVector.SPECIES_PREFERRED;
        var sum = IntVector.zero(species);
        return sum.add(1).add(2).add(3).add(4).add(5).add(6).add(7).add(8)
                .add(9).add(10).add(11).add(12).add(13).add(14).add(15).add(16)
                .add(17).add(18).add(19).add(20).add(21).add(22).add(23).add(24)
                .add(25).add(26).add(27).add(28).add(29).add(30).add(31).add(32)
                .add(1).add(2).add(3).add(4).add(5).add(6).add(7).add(8)
                .add(9).add(10).add(11).add(12).add(13).add(14).add(15).add(16)
                .add(17).add(18).add(19).add(20).add(21).add(22).add(23).add(24)
                .add(25).add(26).add(27).add(28).add(29).add(30).add(31).add(32)
                .reinterpretAsLongs()
                .lane(0);
    }

provides the following result:

    Before:
    Benchmark                     Mode  Cnt   Score   Error  Units
    VectorReplicate.broadcastCon  avgt    5  16.417 ± 0.515  ns/op

    After:
    Benchmark                     Mode  Cnt   Score   Error  Units
    VectorReplicate.broadcastCon  avgt    5  13.851 ± 0.154  ns/op

The constant table size decreases from 1024 bytes to 128 bytes, which is much more manageable. The throughput improvement mostly comes from the vector being rematerialized instead of being spilt on the stack.

I have not been able to observe performance gain regarding bypass delay, which is expected as according to "Agner's optimisation manual on the micro architecture of Intel, AMD and VIA CPUs", Intel CPUs since Skylake seem to have only a few such delays.

Thank you very much.

**Comment by merykitty** [comment]
I have come back to this PR with some modifications and added a benchmark for this patch. The description is also modified to better present the objective of this patch and show the results. Please have some reviews, thank you very much.

**Review by vnkozlov — COMMENTED** [review]
I submitted testing.

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 1]
```diff
@@ -4363,11 +4385,15 @@ void MacroAssembler::_verify_oop(Register reg, const char* s, const char* file,
 }
 
 void MacroAssembler::vallones(XMMRegister dst, int vector_len) {
-  if (UseAVX > 2 && (vector_len == Assembler::AVX_512bit || VM_Version::supports_avx512vl())) {
+  // vpcmpeqd has special dependency treatment so it should be preferred to vpternlogd
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:4388]: Comment is not clear.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 2]
```diff
@@ -1641,23 +1641,34 @@ void C2_MacroAssembler::load_vector_mask(KRegister dst, XMMRegister src, XMMRegi
...
+    }
+  } else {
+    switch (vlen_in_bytes) {
+      case 4:  movflt(dst, src); break;
+      case 8:  movdbl(dst, src); break;
+      case 16: movups(dst, src); break;
+      case 32: vmovups(dst, src, Assembler::AVX_256bit); break;
+      case 64: vmovups(dst, src, Assembler::AVX_512bit); break;
+      default: ShouldNotReachHere();
+    }
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:1662]: Vector Load/store from memory happens from dedicated ports, can you elaborate why this change will benefit.

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:1662]: > Vector Load/store from memory happens from dedicated ports, can you elaborate why this change will benefit.

Above reference to section 3.5.5.2 also states that FP loads adds another cycle of latency,  but saving the cycles penalty due to bypass b/w FP and SIMD domains still holds good. So may be for load there is no pressing need and existing load vector handling can be kept as it is.

Overall savings from constant table size reductions are very impressive.  Thanks.

**Comment by merykitty** [comment]
@jatin-bhateja Thanks a lot for your comments, I have addressed them in the last commit

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp (hunk 2 above)
Comment (merykitty) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: Thanks for your sharing, I have reverted the change here

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp (hunk 1 above)
Comment (merykitty) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp]: I have remeasured the statement, it seems that only the non-vex encoding version receives the special dependency treatment, so I reverted this change and add a comment for clarification.

The optimisation can be found noticed in [The microarchitecture of Intel, AMD and VIA CPUs: An optimization guide for assembly programmers and compiler makers](https://www.agner.org/optimize/) on several architectures such as in section 9.8 (Register allocation and renaming in Sandy Bridge and Ivy Bridge pipeline).

I have performed measurements on uica.uops.info . While this sequence gives 1.37 cycles/iteration on Skylake and Icelake

    pcmpeqd xmm0, xmm0
    paddd xmm0, xmm1
    paddd xmm0, xmm1
    paddd xmm0, xmm1

This version has the throughput of 4 cycles/iteration

    vpcmpeqd xmm0, xmm0, xmm0
    vpaddd xmm0, xmm1, xmm0
    vpaddd xmm0, xmm1, xmm0
    vpaddd xmm0, xmm1, xmm0

Which indicates the `vpcmpeqd` failing to break dependencies on `xmm0` as opposed to the `pcmpeqd` instruction.

Thanks.

**Comment by vnkozlov** [comment]
The testing of version 07 got failure when run vector tests with `-XX:UseAVX=0 -XX:UseSSE=2`:
```
# Internal Error (/workspace/open/src/hotspot/share/opto/constantTable.cpp:217), pid=2750036, tid=2750067
#  assert((constant_addr - _masm.code()->consts()->start()) == con.offset()) failed: must be: 8 == 0

Current CompileTask:
C2:    287   29 %  b        compiler.codegen.TestByteVect::test_ci @ 2 (20 bytes)

Stack: [0x00007f7abf144000,0x00007f7abf245000],  sp=0x00007f7abf23fa30,  free space=1006k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0xb731c8]  ConstantTable::emit(CodeBuffer&) const+0x1c8
V  [libjvm.so+0x17c3673]  PhaseOutput::fill_buffer(CodeBuffer*, unsigned int*)+0x293
V  [libjvm.so+0xb191bb]  Compile::Code_Gen()+0x42b
V  [libjvm.so+0xb1e899]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x1699
```

and 

```
#  Internal Error (/workspace/open/src/hotspot/cpu/x86/assembler_x86.cpp:5095), pid=1431469, tid=1431493
#  Error: assert(VM_Version::supports_ssse3()) failed

Current CompileTask:
C2:    468  240 %  b  4       java.util.Arrays::fill @ 5 (21 bytes)

Stack: [0x00007fdecd422000,0x00007fdecd523000],  sp=0x00007fdecd51d8c0,  free space=1006k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x73079c]  Assembler::pshufb(XMMRegisterImpl*, XMMRegisterImpl*)+0x13c
V  [libjvm.so+0x4005d1]  ReplB_regNode::emit(CodeBuffer&, PhaseRegAlloc*) const+0x1a1
V  [libjvm.so+0x17be04e]  PhaseOutput::scratch_emit_size(Node const*)+0x45e
V  [libjvm.so+0x17b4548]  PhaseOutput::shorten_branches(unsigned int*)+0x2d8
V  [libjvm.so+0x17c6faa]  PhaseOutput::Output()+0xcfa
V  [libjvm.so+0xb191bb]  Compile::Code_Gen()+0x42b
V  [libjvm.so+0xb1e899]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x1699
```

**Comment by merykitty** [comment]
@vnkozlov I have fixed those errors in the last commits. The second one is due to `pshufb` being supported only on ssse3 machines. And the first one is because the constant table itself is not aligned enough given that currently, it is only aligned at 8 bytes. I chose to avoid the problem and only emit constants requiring at most 8 bytes of alignment as this patch has already touched many areas. A proper solution would be in a separate issue. What do you think?

Thanks a lot.

https://github.com/openjdk/jdk/blob/2bd90c2149bfee4b045c8f376e8bcdf4420ccb5d/src/hotspot/share/asm/codeBuffer.hpp#L742

**Comment by vnkozlov** [comment]
> @vnkozlov I have fixed those errors in the last commits. The second one is due to `pshufb` being supported only on ssse3 machines. And the first one is because the constant table itself is not aligned enough given that currently, it is only aligned at 8 bytes. I chose to avoid the problem and only emit constants requiring at most 8 bytes of alignment as this patch has already touched many areas. A proper solution would be in a separate issue. What do you think?
> 
I agree with doing in separate changes. And I will start new testing.

**Comment by vnkozlov** [comment]
Got new failure (and testing still running). Test compiler/c2/cr7200264/TestSSE2IntVect.java failed with `-Xcomp`:
```
java.lang.RuntimeException: Unexpected SubVI number: expected 2 >= 4
	at jdk.test.lib.Asserts.fail(Asserts.java:594)
	at jdk.test.lib.Asserts.assertGreaterThanOrEqual(Asserts.java:288)
	at jdk.test.lib.Asserts.assertGTE(Asserts.java:259)
	at compiler.c2.cr7200264.TestDriver.verifyVectorizationNumber(TestDriver.java:65)
	at compiler.c2.cr7200264.TestDriver.run(TestDriver.java:43)
	at compiler.c2.cr7200264.TestSSE2IntVect.main(TestSSE2IntVect.java:48)
```

**Comment by merykitty** [comment]
It does not seem related as this patch has effects only after matching so it should not change the IR graph of the compilations

**Review by vnkozlov — APPROVED** [review]
I verified that the latest failure I posted is not related to these changes. There were no other failures. Approved.

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp (hunk 1 above)
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp]: Both the above JIT sequences have true dependency chain,  there is no scope of any additional architecture imposed false dependency doing any further perf degradation for which we use dep-breaking idioms.

Comment (merykitty) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp]: I'm sorry I don't quite understand what do you mean here, what I meant is that while `pcmpeqd xmmk, xmmk` is a dep-breaking idiom, `vpcmpeqd xmmk, xmmk, xmmk` seems to not be. As a result, I reverted that change and in this context, the only change is I added a branch for non-AVX machines. Please have a review for this patch. Thank you very much.

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp]: Yes, its a valid one-idiom and as per section E.1.2 of [X86 Optimization manual](https://cdrdv2.intel.com/v1/dl/getContent/671488)  such idioms are resolved by renamer and does not reach execution ports. 

I faintly remember that there was a subtle difference b/w handling of zeroing/one idioms on certain targets where in some cases one-idioms still go beyond renamer. But, we can keep this change of your since even if all-one idiom (vpcmpeqd) reach execution port, latency wise it same a vpternlog over 256 bit vector.

File: src/hotspot/share/opto/machnode.cpp [hunk 3]
```diff
@@ -474,14 +474,15 @@ bool MachNode::rematerialize() const {
   }
 
   // Stretching lots of inputs - don't do it.
-  if (req() > 2) {
+  // A MachContant has the last input being the constant base
+  if (req() > (is_MachConstant() ? 3U : 2U)) {
```
[the remark this answers is not in the document]
Comment (jatin-bhateja) [reply src/hotspot/share/opto/machnode.cpp:478]: Saving a spill at the cost of re-materialization using a comparatively cheaper instruction like add/sub/mul looks better for divD may be costly.

[the remark this answers is not in the document]
Comment (jatin-bhateja) [reply src/hotspot/share/opto/machnode.cpp:478]: There are other machine nodes which just accept constants as a mode, like vround* and vcompu* nodes which will now qualify for rematerlization leading to emitting high cost instructions.

[the remark this answers is not in the document]
Comment (jatin-bhateja) [reply src/hotspot/share/opto/machnode.cpp:478]: I think we should have a rough cost model here and not just basing it purely over connectivity of the node, or for the time being you can remove this change ?

File: src/hotspot/cpu/x86/x86.ad [hunk 4]
```diff
@@ -4107,103 +4107,96 @@ instruct scatter_masked(memory mem, vec src, vec idx, kReg mask, kReg ktmp, rReg
...
+    __ vpbroadcastb($dst$$XMMRegister, $mem$$Address, vlen_enc);
   %}
   ins_pipe( pipe_slow );
 %}
 
 // ====================ReplicateS=======================================
 
-instruct ReplS_reg(vec dst, rRegI src) %{
+instruct vReplS_reg(vec dst, rRegI src) %{
+  predicate(UseAVX >= 2);
```
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad:4159]: Aligning the predicate of the reg and the mem version allows the adlc parser to recognise their relationship and during register allocation can substitute a reg operation with a spilt operand with its corresponding mem node.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 5]
```diff
@@ -1643,12 +1643,12 @@ void C2_MacroAssembler::load_vector_mask(KRegister dst, XMMRegister src, XMMRegi
...
-  case 16: movdqu(dst, src);  break;
-  case 32: vmovdqu(dst, src); break;
-  case 64: evmovdquq(dst, src, Assembler::AVX_512bit); break;
-  default: ShouldNotReachHere();
+    case 4:  movdl(dst, src);   break;
+    case 8:  movq(dst, src);    break;
+    case 16: movdqu(dst, src);  break;
+    case 32: vmovdqu(dst, src); break;
+    case 64: evmovdqul(dst, src, Assembler::AVX_512bit); break;
+    default: ShouldNotReachHere();
```
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:1651]: Since I added the method `C2_MacroAssembler::load_constant_vector` near here anyway I think this style change can be kept.

File: src/hotspot/cpu/x86/x86.ad [hunk 6]
```diff
@@ -4238,25 +4226,39 @@ instruct ReplI_mem(vec dst, memory mem) %{
...
+    if (VM_Version::supports_avx()) {
+      __ vbroadcastss($dst$$XMMRegister, addr, vlen_enc);
+    } else if (VM_Version::supports_sse3()) {
+      __ movddup($dst$$XMMRegister, addr);
+    } else {
+      __ movq($dst$$XMMRegister, addr);
+      if (Matcher::vector_length_in_bytes(this) >= 16) {
+        __ punpcklqdq($dst$$XMMRegister, $dst$$XMMRegister);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad]: Done

File: src/hotspot/cpu/x86/x86.ad [hunk 7]
```diff
@@ -4238,25 +4226,39 @@ instruct ReplI_mem(vec dst, memory mem) %{
...
-    InternalAddress addr = $constantaddress(T_INT, vreplicate_imm(T_INT, $con$$constant, Matcher::vector_length(this)));
-    __ load_vector($dst$$XMMRegister, addr, Matcher::vector_length_in_bytes(this));
+    InternalAddress addr = $constantaddress(Matcher::vector_element_basic_type(this),
+        vreplicate_imm(Matcher::vector_element_basic_type(this), $con$$constant,
+            (VM_Version::supports_sse3() ? (VM_Version::supports_avx() ? 4 : 8) : 8) /
+                type2aelembytes(Matcher::vector_element_basic_type(this))));
+    BasicType bt = Matcher::vector_element_basic_type(this);
+    int vlen_enc = vector_length_encoding(this);
+    if (VM_Version::supports_avx()) {
+      __ vbroadcastss($dst$$XMMRegister, addr, vlen_enc);
```
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad]: Got it

File: src/hotspot/share/opto/machnode.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/share/opto/machnode.cpp:478]: A node is decided to prefer rematerialising to spilling has to satisfy that:

- The node is not explicitly said to be expensive, `divD` and `divF` fails at this stage.
- The node declaration only contains simple register rules (explicit or implicit DEF dst and USE src), `vround` fails this because it has temp register, `cmpF_imm` and `cmpD_imm` fail this because they kill flags.
- This method we are at agrees with the rematerialising.

I have looked at all instances where `constantaddress` is used and found no node where accidental rematerialisation is inefficient.

**Comment by merykitty** [comment]
@jatin-bhateja Thanks a lot for your comments, I have addressed those in the last commit.
@vnkozlov Thanks very much for the review and testing.

File: src/hotspot/cpu/x86/x86.ad (hunk 4 above)
[the remark this answers is not in the document]
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/x86.ad:4159]: May be a misplaced comment, what I meant was to collapse patterns if number and register class of operands comply.

**Review by jatin-bhateja — APPROVED** [review]

File: src/hotspot/share/opto/machnode.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jatin-bhateja) [reply src/hotspot/share/opto/machnode.cpp:478]: Thanks for your explanations, I agree.

**Comment by merykitty** [comment]
Thanks for your reviews. Does this PR need another run through the tests?

**Review by vnkozlov — COMMENTED** [review]
I submitted new testing for version 012. Last update was not simple.

**Review by vnkozlov — APPROVED** [review]
Testing of version 12 passed.

**Comment by merykitty** [comment]
May I have this PR sponsored, please? Thanks a lot for your help.
